planner/core/memtable_predicate_extractor.go

```diff
@@ -466,3 +469,95 @@ func (e *ClusterLogTableExtractor) Extract(
 	e.Patterns = patterns
 	return remained
 }
+
+// MetricTableExtractor is used to extract some predicates of metric_schema tables.
+type MetricTableExtractor struct {
+	extractHelper
+	// SkipRequest means the where clause always false, we don't need to request any component
+	SkipRequest bool
+	// StartTime represents the beginning time of metric data.
+	StartTime int64
+	// EndTime represents the ending time of metric data.
+	EndTime int64
+	// LabelConditions represents the label conditions of metric data.
+	LabelConditions map[string]set.StringSet
+	quantile        set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *MetricTableExtractor) Extract(
+	ctx sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) []expression.Expression {
+	// Extract the `quantile` columns
+	remained, skipRequest, quantile := e.extractCol(schema, names, predicates, "quantile", true)
+	e.SkipRequest = skipRequest
+	if e.SkipRequest {
+		return nil
+	}
+	e.quantile = quantile
+
+	// Extract the `time` columns
+	remained, startTime, endTime := e.extractTimeRange(ctx, schema, names, remained, "time")
+	if endTime == 0 {
+		endTime = math.MaxInt64
+	}
+	e.StartTime = startTime
+	e.EndTime = endTime
+	e.SkipRequest = startTime > endTime
+
+	if e.SkipRequest {
+		return nil
+	}
+
+	// Extract the label columns.
+	for _, name := range names {
+		switch name.ColName.L {
+		case "quantile", "time", "value":
+			continue
+		}
+		var values set.StringSet
+		remained, skipRequest, values = e.extractCol(schema, names, remained, name.ColName.L, false)
+		if skipRequest {
+			e.SkipRequest = skipRequest
+			return nil
+		}
+		if len(values) == 0 {
+			continue
+		}
+		if e.LabelConditions == nil {
+			e.LabelConditions = make(map[string]set.StringSet)
+		}
+		e.LabelConditions[name.ColName.L] = values
+	}
+	return remained
+}
+
+// GetQuantile gets the quantile of metric query.
+func (e *MetricTableExtractor) GetQuantile() (quantile float64, err error) {
+	if len(e.quantile) == 0 {
+		return 0, nil
+	}
+	if len(e.quantile) > 1 {
+		return 0, errors.New("query metric data not support specified multiple quantile")
+	}
+	for k := range e.quantile {
+		quantile, err = strconv.ParseFloat(k, 64)
+		return
+	}
+	return 0, nil
+}
+
+// GetQueryRangeTime gets the metric query time range.
+func (e *MetricTableExtractor) GetQueryRangeTime(sctx sessionctx.Context) (start, end time.Time, step time.Duration) {
+	return e.convertToTime(e.StartTime), e.convertToTime(e.EndTime), time.Second * time.Duration(sctx.GetSessionVars().MetricSchemaStep)
+}
+
+func (e *MetricTableExtractor) convertToTime(t int64) time.Time {
+	if t == 0 || t == math.MaxInt64 {
+		return time.Now()
```

Comment:
What if startTime=0, endTime=now-1h?
You changed it to [now, now-1h].

Reply:
Yes, but it will got error when query.
```sql
mysql>desc select * from query_duration where `Quantile` in ( 0.99) and sql_type="internal" and time < "2019-12-23 16:10:13" and instance="127.0.0.1:10080";
+----------------+----------+------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id             | count    | task | operator info                                                                                                                                                                                                                          |
+----------------+----------+------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_5 | 10000.00 | root | PromQL:histogram_quantile(0.99, sum(rate(tidb_server_handle_query_duration_seconds_bucket{instance="127.0.0.1:10080",sql_type="internal"}[60s])) by (le)), start_time:2019-12-23 16:20:47.559, end_time:2019-12-23 16:10:12, step:1m0s |
+----------------+----------+------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set
Time: 0.009s
mysql>select * from query_duration where `Quantile` in ( 0.99) and sql_type="internal" and time < "2019-12-23 16:10:13" and instance="127.0.0.1:10080";
(1105, u'bad_data: end timestamp must not be before start time')
```

Reply:
I mean, it shouldn't return an error.
When user queries [0, 16:10], you can change it to [13:40, 16:10], not [16:20, 16:10].

Reply:
addressed. But When the user queries [0, 16:10], it will change it to [16:10, 16:10] now.

Reply:
You can refer to
https://github.com/pingcap/tidb/blob/master/executor/cluster_reader.go#L469

Reply:
ok, done